presences/src/main/resources/public/ts/directives/events-filter-form/events-filter-form.ts

```diff
@@ -0,0 +1,337 @@
+    reasons: Array<Reason>;
+    noReason: Reason;
+    filter: EventsFormFilter;
+    formFilter: EventsFormFilter;
+    structureTimeSlot: IStructureSlot;
+
+    studentsSearch: StudentsSearch;
+    groupsSearch: GroupsSearch;
+
+    constructor(private $scope: IScope,
+                private $location: ILocationService,
+                private $window: IWindowService,
+                private searchService: SearchService,
+                private groupService: GroupService) {
+
+        // On display, load search bars data
+        this.$scope.$watch(() => this.display, async () => {
+            if (this.studentsSearch && this.groupsSearch && this.studentsSearch.structureId &&
+                this.groupsSearch.structureId) {
+                this.studentsSearch.setSelectedStudents(this.filter.students);
+                this.groupsSearch.setSelectedGroups(this.filter.classes);
+            }
+            if (!this.formFilter) {
+                await this.initData();
+            }
+        });
+
+        this.noReason = {
+            id: 0,
+            label: this.translate('presences.absence.no.reason'),
+            structure_id: '',
+            comment: '',
+            default: false,
+            proving: false,
+            group: false
+        } as Reason;
+
+        this.$parent = undefined;
+    }
+
+    translate = (key: string): string => idiom.translate(key);
+
+    async $onInit(): Promise<void> {
+        this.display = false;
+        await this.initData();
+    }
+
+    async initData(): Promise<void> {
+        this.structureTimeSlot = {} as IStructureSlot;
+        let formFilterPref: EventsFormFilter = this.absencesOnly ?
+            await Me.preference(PresencesPreferenceUtils.PREFERENCE_KEYS.PRESENCE_PLANNED_ABSENCES_FILTER) :
+            await Me.preference(PresencesPreferenceUtils.PREFERENCE_KEYS.PRESENCE_EVENT_LIST_FILTER);
+        formFilterPref = formFilterPref ? formFilterPref[window.structure.id] : null;
+        this.formFilter = formFilterPref ? formFilterPref : JSON.parse(JSON.stringify(this.filter));
+        this.studentsSearch = new StudentsSearch(window.structure.id, this.searchService);
+        this.groupsSearch = new GroupsSearch(window.structure.id, this.searchService, this.groupService);
+        await Promise.all([this.loadReasons(), this.getStructureTimeSlots()]);
+        if (this.formFilter.timeslots && this.formFilter.timeslots.start && this.formFilter.timeslots.end) {
+            this.formFilter.timeslots = {
+                start: this.structureTimeSlot.slots.find((slot: ITimeSlot) =>
+                    slot._id === this.formFilter.timeslots.start._id),
+                end: this.structureTimeSlot.slots.find((slot: ITimeSlot) =>
+                    slot._id === this.formFilter.timeslots.end._id)
+            };
+        }
+    }
+
+    async loadReasons(): Promise<void> {
+        this.reasons = await reasonService.getReasons(window.structure.id, REASON_TYPE_ID.ALL);
```

Comment:
Si tu fail, ton this.reasons sera undefined donc une erreur cannot read property of undefined